### PR TITLE
Minimal runtime pthreads

### PR DESCRIPTION
**Pull request by juj**
This adds support for multithreading with MINIMAL_RUNTIME, in both regular build mode and MODULARIZEd build mode.

### CODE REVIEW

**Comment by juj** [comment]
Brought this PR up to latest and all tests pass. This would be great to see a review and get landed soon, it'll bump up MINIMAL_RUNTIME features with threading support!

File: src/library.js [hunk 1]
```diff
@@ -1335,17 +1341,6 @@ LibraryManager.library = {
...
-  $establishStackSpace__asm: true,
-  $establishStackSpace__sig: 'vii',
-  $establishStackSpace: function(stackBase, stackMax) {
-    stackBase = stackBase|0;
-    stackMax = stackMax|0;
-    STACKTOP = stackBase;
-    STACK_MAX = stackMax;
-  },
-#endif
-
```
Comment (juj) [inline src/library.js]: This deletion is not strictly related to MINIMAL_RUNTIME multithreading support, but while doing this PR I realized this whole function is redundant, so simplified the code.

File: emscripten.py [hunk 2]
```diff
@@ -1887,9 +1885,13 @@ def create_runtime_funcs_asmjs(exports, metadata):
...
 ''' % stack_check]
 
+  if shared.Settings.MINIMAL_RUNTIME:
+    # MINIMAL_RUNTIME moves stack functions to library.
+    funcs = []
+
   if shared.Settings.USE_PTHREADS:
     funcs.append('''
-function establishStackSpace(stackBase, stackMax) {
+function asmJsEstablishStackFrame(stackBase, stackMax) {
```
Comment (juj) [inline emscripten.py]: Renamed this to highlight the fact that this functionality is only needed for old fastcomp threading support.

File: src/library_pthread.js [hunk 3]
```diff
@@ -400,17 +402,23 @@ var LibraryPThread = {
         'buffer': HEAPU8.buffer,
         'asmJsUrlOrBlob': Module["asmJsUrlOrBlob"],
 #endif
+#if !MINIMAL_RUNTIME
         'DYNAMIC_BASE': DYNAMIC_BASE,
+#endif
```
Comment (juj) [inline src/library_pthread.js]: In MINIMAL_RUNTIME the `DYNAMIC_BASE` variable is not a thing.

File: src/library_pthread.js [hunk 4]
```diff
@@ -400,17 +402,23 @@ var LibraryPThread = {
...
         'DYNAMICTOP_PTR': DYNAMICTOP_PTR
       });
     },
 
     // Creates a new web Worker and places it in the unused worker pool to wait for its use.
     allocateUnusedWorker: function() {
+#if MINIMAL_RUNTIME
+      var pthreadMainJs = Module['worker'];
```
Comment (juj) [inline src/library_pthread.js]: This use of `Module['worker']` is probably temporary/internal, and will go away later as the `src/worker.js` refactor happens.

File: src/library_pthread.js [hunk 5]
```diff
@@ -769,17 +777,30 @@ var LibraryPThread = {
...
+    if (ENVIRONMENT_IS_NODE) return;
 #endif
+
+    if (ENVIRONMENT_IS_PTHREAD) return; // Blocking in a pthread is fine.
+
+    warnOnce('Blocking on the main thread is very dangerous, see https://emscripten.org/docs/porting/pthreads.html#blocking-on-the-main-browser-thread');
 #if !ALLOW_BLOCKING_ON_MAIN_THREAD
     abort('Blocking on the main thread is not allowed by default. See https://emscripten.org/docs/porting/pthreads.html#blocking-on-the-main-browser-thread');
+#endif
+
```
Comment (juj) [inline src/library_pthread.js]: In MINIMAL_RUNTIME we do not want to pay code size for `emscripten_check_blocking_allowed` type of checks, so this is not generated there.

File: src/memoryprofiler.js [hunk 6]
```diff
@@ -597,3 +599,5 @@ var emscriptenMemoryProfiler = {
 function memoryprofiler_add_hooks() { emscriptenMemoryProfiler.initialize(); }
 
 if (typeof Module !== 'undefined' && typeof document !== 'undefined' && typeof window !== 'undefined' && typeof process === 'undefined') emscriptenMemoryProfiler.initialize();
+
+#endif
```
Comment (juj) [inline src/memoryprofiler.js]: Moving this #ifdef was done to be more idiomatic with include guards (the file itself guards, rather than the includer of that file)

File: tests/pthread/call_async_on_main_thread.js [hunk 7]
```diff
@@ -3,7 +3,7 @@ mergeInto(LibraryManager.library, {
   report_result__proxy: 'async',
   report_result__sig: 'viii',
   report_result: function(param1, param2, param3) {
-    if (ENVIRONMENT_IS_WORKER) {
+    if (typeof ENVIRONMENT_IS_WORKER !== 'undefined' && ENVIRONMENT_IS_WORKER) {
```
Comment (juj) [inline tests/pthread/call_async_on_main_thread.js]: This test change and all the test changes below are to make pthreads tests pass in MINIMAL_RUNTIME. It enables a workflow where I locally edit `src/settings.js` directly to change `var MINIMAL_RUNTIME = 1;` by default to run the whole browser.test_pthread* test suite in MINIMAL_RUNTIME, even though we don't duplicate all browser tests twice for MINIMAL_RUNTIME and non-MINIMAL_RUNTIME cases (only select tests are duplicated)

File: emcc.py [hunk 8]
```diff
@@ -3427,8 +3434,7 @@ def module_export_name_substitution():
     src = src.replace(shared.JS.module_export_name_substitution_pattern, replacement)
     # For Node.js and other shell environments, create an unminified Module object so that
     # loading external .asm.js file that assigns to Module['asm'] works even when Closure is used.
-    if shared.Settings.MINIMAL_RUNTIME and (shared.Settings.target_environment_may_be('node') or
-                                            shared.Settings.target_environment_may_be('shell')):
+    if shared.Settings.MINIMAL_RUNTIME and not shared.Settings.MODULARIZE_INSTANCE and (shared.Settings.target_environment_may_be('node') or shared.Settings.target_environment_may_be('shell')):
```
Comment (kripken) [inline emcc.py]: why does modularize_instance matter here? (and, why not modularize?)

File: src/library_pthread.js [hunk 9]
```diff
@@ -769,17 +777,30 @@ var LibraryPThread = {
...
   },
 
+#if MINIMAL_RUNTIME
+  emscripten_check_blocking_allowed__deps: ['$warnOnce'],
+#endif
   emscripten_check_blocking_allowed: function() {
-#if ASSERTIONS
-    assert(ENVIRONMENT_IS_WEB);
-    warnOnce('Blocking on the main thread is very dangerous, see https://emscripten.org/docs/porting/pthreads.html#blocking-on-the-main-browser-thread');
+#if ASSERTIONS || IN_TEST_HARNESS || !MINIMAL_RUNTIME || !ALLOW_BLOCKING_ON_MAIN_THREAD
```
Comment (kripken) [inline src/library_pthread.js]: I don't understand why `!MINIMAL_RUNTIME` is here? This code is already behind `ASSERTIONS`, so it won't be in an optimized build with the minimal runtime or not? I think it's useful in assertion builds to have maximal checks, regardless of runtime.

File: src/modules.js [hunk 10]
```diff
@@ -86,6 +86,10 @@ var LibraryManager = {
       libraries.push('library_browser.js');
     }
 
+    if (USE_PTHREADS) { // TODO: Currently WebGL proxying makes pthreads library depend on WebGL.
+      libraries.push('library_webgl.js');
+    }
```
Comment (kripken) [inline src/modules.js]: is this a fix unrelated to this PR?

(we don't deduplicate this list, so this is extra work, and potentially buggy if we have global state - we should deduplicate it before doing this, probably)

File: src/worker.js [hunk 11]
```diff
@@ -81,23 +62,54 @@ var wasmOffsetData;
 this.onmessage = function(e) {
   try {
     if (e.data.cmd === 'load') { // Preload command that is called once per worker to parse and load the Emscripten code.
+#if MINIMAL_RUNTIME
+#if MODULARIZE_INSTANCE
+      var imports = Module;
+#else
+      var imports = {};
```
Comment (kripken) [inline src/worker.js]: how does this empty imports object work? is it filled later? (seems not?) do we not need imports somehow?

File: src/worker.js [hunk 12]
```diff
@@ -81,23 +62,54 @@ var wasmOffsetData;
...
+#if USES_DYNAMIC_ALLOC
+      {{{ makeAsmImportsAccessInPthread('DYNAMICTOP_PTR') }}} = e.data.DYNAMICTOP_PTR;
+#endif
 
 #if WASM
       // Module and memory were sent from main thread
+#if MINIMAL_RUNTIME
+#if MODULARIZE
+      imports['wasm'] = e.data.wasmModule; // Pass the shared Wasm module in an imports object for the MODULARIZEd build.
+#else
```
Comment (kripken) [inline src/worker.js]: can't `makeAsmImportsAccessInPthread` avoid this else?

File: src/worker.js [hunk 13]
```diff
@@ -81,23 +62,54 @@ var wasmOffsetData;
...
+#endif
+
       // Initialize the global "process"-wide fields:
+#if !MINIMAL_RUNTIME
       Module['DYNAMIC_BASE'] = e.data.DYNAMIC_BASE;
-      Module['DYNAMICTOP_PTR'] = e.data.DYNAMICTOP_PTR;
+#endif
+
+#if USES_DYNAMIC_ALLOC
+      {{{ makeAsmImportsAccessInPthread('DYNAMICTOP_PTR') }}} = e.data.DYNAMICTOP_PTR;
```
Comment (kripken) [inline src/worker.js]: this is new - is it only needed for minimal_runtime?

File: src/worker.js [hunk 14]
```diff
@@ -131,14 +143,20 @@ this.onmessage = function(e) {
...
+#else
       Module = {{{ EXPORT_NAME }}}(Module);
 #endif
-      PThread = Module['PThread'];
-      HEAPU32 = Module['HEAPU32'];
-      postMessage({ cmd: 'loaded' });
+#endif
+
+#if !MINIMAL_RUNTIME || !WASM
+      postMessage({ 'cmd': 'loaded' });
```
Comment (kripken) [inline src/worker.js]: how does minimal runtime avoid the need for `loaded`? and non-WASM?

File: src/worker.js [hunk 15]
```diff
@@ -131,14 +143,20 @@ this.onmessage = function(e) {
         URL.revokeObjectURL(objectUrl);
       }
 #if MODULARIZE && !MODULARIZE_INSTANCE
+#if MINIMAL_RUNTIME
+      Module = {{{ EXPORT_NAME }}}(imports);
+#else
       Module = {{{ EXPORT_NAME }}}(Module);
 #endif
-      PThread = Module['PThread'];
```
Comment (kripken) [inline src/worker.js]: isn't it shorter to assign to `PThread` once instead of doing `Module['PThread']` many times as you changed it to?

File: src/worker.js [hunk 16]
```diff
@@ -197,23 +222,36 @@ this.onmessage = function(e) {
 #if STACK_OVERFLOW_CHECK
         Module['checkStackCookie']();
 #endif
+#if !MINIMAL_RUNTIME // In MINIMAL_RUNTIME the noExitRuntime concept does not apply to pthreads. To exit a pthread with live runtime, use the function emscripten_unwind_to_js_event_loop() in the pthread body.
```
Comment (kripken) [inline src/worker.js]: this sort of difference seems dangerously confusing. is it well documented? Are there assertions that can catch the corner cases?

File: tests/pthread/call_sync_on_main_thread.c [hunk 17]
```diff
@@ -19,15 +20,15 @@ extern int receivesAndReturnsAnInteger(int i);
 
 int main()
 {
-	char dst[256];
-	char name[7] = "resize";
-	getDomElementParentInnerHTML(name, dst, sizeof(dst));
+	char name[7] = "body";
+	const char *dst = getDomElementContents(name);
 	memset(name, 0, sizeof(name)); // Try to uncover if there might be a race condition and above line was not synchronously processed, and we could take name string away.
```
Comment (kripken) [inline tests/pthread/call_sync_on_main_thread.c]: are the changes to this file necessary? (comment above said they were, but I don't see how?)

File: tests/pthread/test_pthread_mutex.cpp [hunk 18]
```diff
@@ -107,7 +108,7 @@ int main()
 		for(int i = 0; i < NUM_THREADS; ++i)
 			CreateThread(i, threadNum++);
 
-		emscripten_set_main_loop(WaitToJoin, 0, 0);
+		emscripten_set_timeout_loop(WaitToJoin, 100, 0);
 	} else {
```
Comment (kripken) [inline tests/pthread/test_pthread_mutex.cpp]: same question for this file - why doesn't it work in minimal_runtime without these?

File: emcc.py (hunk 8 above)
Comment (juj) [reply emcc.py]: MODULARIZE_INSTANCE creates the Module on the next line, emitting

```
if(typeof Module==="undefined"){var Module={};}
var Module = (
```
would be redundant for code size.

File: src/library_pthread.js (hunk 9 above)
Comment (juj) [reply src/library_pthread.js]: Agreed, let me drop that.

Comment (juj) [reply src/library_pthread.js]: Oh, that is a `|| !MINIMAL_RUNTIME` statement and not a ` && !MINIMAL_RUNTIME` statement. The `|| !MINIMAL_RUNTIME` statement is here so that regular runtime will get the non-assertion release builds also to contain the allow blocking checks.

File: src/modules.js (hunk 10 above)
Comment (juj) [reply src/modules.js]: This fix is not unrelated - in MINIMAL_RUNTIME  and STRICT we don't autolink libraries, and that would cause pthreads builds to fail at missing GL libraries. Added deduplication.

File: src/worker.js (hunk 11 above)
Comment (juj) [reply src/worker.js]: The imports are filled from `makeAsmImportsAccessInPthread()` function.

File: src/worker.js (hunk 12 above)
Comment (juj) [reply src/worker.js]: Oh right, good catch.

File: src/worker.js (hunk 13 above)
Comment (juj) [reply src/worker.js]: Hmm, not sure what you mean? Importing DYNAMICTOP_PTR is needed for both regular and minimal runtime?

File: src/worker.js (hunk 14 above)
Comment (juj) [reply src/worker.js]: Good point: MINIMAL_RUNTIME always uses async wasm compilation - even in pthreads. So it cannot signal 'loaded' here yet, but it will do that signaling inside the wasm compilation .then() handler. Added a comment to clarify this.

File: src/worker.js (hunk 15 above)
Comment (juj) [reply src/worker.js]: It is, but I wanted to avoid the number of global variables that would collide in meaning: having PThread and HEAPU32 here, but then loading the global JS script with the same vars can be confusing.

Proper code size optimizations will follow with the refactor to remove all of these Module['x'] exports later by moving the contents of src/worker.js into the main JS file (and keeping src/worker.js a minimal stub that only loads the main JS) - that will allow closure minification of all of this.

File: src/worker.js (hunk 16 above)
Comment (juj) [reply src/worker.js]: Yes, there should be clear diagnostics messages that were added in #10174.

File: tests/pthread/call_sync_on_main_thread.c (hunk 17 above)
Comment (juj) [reply tests/pthread/call_sync_on_main_thread.c]: The old test asserted the existence of an inner HTML text node with text contents "resize" in it (that was the [ ] resize canvas checkbox in default shell). In MINIMAL_RUNTIME shell such checkbox does not exist, so I replaced it with a string snippet `<canvas` that can be found in both shells.

File: tests/pthread/test_pthread_mutex.cpp (hunk 18 above)
Comment (juj) [reply tests/pthread/test_pthread_mutex.cpp]: MINIMAL_RUNTIME does not support library_browser.js because the $Browser object does not DCE well if linked in. This test was not looking to exercise `emscripten_set_main_loop`, so changed to `emscripten_set_timeout_loop` for the same effect.

File: src/worker.js (hunk 12 above)
Comment (juj) [reply src/worker.js]: Err actually, this is a different case since the wasm object flows directly from HTML, but other variables are in global scope in non-MODULARIZE builds.

**Comment by juj** [comment]
Ping, any chance this could get a second review pass?
